pyiron_atomistics/lammps/potentials.py

```diff
@@ -0,0 +1,414 @@
+import numpy as np
+import warnings
+
+
+class LammpsPotentials:
+    def __new__(cls, *args, **kwargs):
+        obj = super().__new__(cls)
+        cls._df = None
+        return obj
+
+    def copy(self):
+        new_pot = LammpsPotentials()
+        new_pot.set_df(self.get_df())
+        return new_pot
+
+    @staticmethod
+    def _harmonize_args(args) -> str:
+        if len(args) == 0:
+            raise ValueError("Chemical elements not specified")
+        if len(args) == 1:
+            args *= 2
+        return list(args)
+
+    @property
+    def model(self) -> str:
+        """Model name (required in pyiron df)"""
+        return "_and_".join(set(self.df.model))
+
+    @property
+    def name(self) -> str:
+        """Potential name (required in pyiron df)"""
+        return "_and_".join(set(self.df.name))
+
+    @property
+    def species(self):
+        """Species defined in the potential"""
+        species = set([ss for s in self.df.interacting_species for ss in s])
+        preset = set(["___".join(s) for s in self.df.preset_species if len(s) > 0])
+        if len(preset) == 0:
+            return list(species)
+        elif len(preset) > 1:
+            raise NotImplementedError(
+                "Currently not possible to have multiple file-based potentials"
+            )
+        preset = list(preset)[0].split("___")
+        return [p for p in preset + list(species - set(preset)) if p != "*"]
+
+    @property
+    def filename(self) -> list:
+        """LAMMPS potential files"""
+        return [f for f in set(self.df.filename) if len(f) > 0]
+
+    @property
+    def citations(self) -> str:
+        """Citations to be included"""
+        return "".join(np.unique([c for c in self.df.citations if len(c) > 0]))
+
+    @property
+    def is_scaled(self) -> bool:
+        """Scaling in pair_style hybrid/scaled and hybrid/overlay (whih is scale=1)"""
+        return "scale" in self.df
+
+    @property
+    def pair_style(self) -> str:
+        """LAMMPS pair_style"""
+        if len(set(self.df.pair_style)) == 1:
+            pair_style = "pair_style " + list(set(self.df.pair_style))[0]
+            if np.max(self.df.cutoff) > 0:
+                pair_style += f" {np.max(self.df.cutoff)}"
+            return pair_style + "\n"
+        elif "scale" not in self.df:
+            pair_style = "pair_style hybrid"
+        elif all(self.df.scale == 1):
+            pair_style = "pair_style hybrid/overlay"
+        else:
+            pair_style = "pair_style hybrid/scaled"
+        for ii, s in enumerate(self.df[["pair_style", "cutoff"]].values):
+            if pair_style.startswith("pair_style hybrid/scaled"):
+                pair_style += f" {self.df.iloc[ii].scale}"
+            pair_style += f" {s[0]}"
+            if s[1] > 0:
+                pair_style += f" {s[1]}"
+        return pair_style + "\n"
+
+    @property
+    def pair_coeff(self) -> list:
+        """LAMMPS pair_coeff"""
+
+        class PairCoeff:
```

Comment:
Why do you define the `PairCoeff` inside the property? This way it is always refefinded on accessing the property.

Reply:
Yeah I was not very sure about how to write it properly. Initially, there was just one property function `pair_coeff` (the one you referred to), but then I realized that it was getting too long. In my attempt to refactor the function, I decided to make a separate class for this purpose, and since it's only used for `pair_coeff`, I put it there inside, to make people easily understand that it's not to be used anywhere else. It's possible that it's not the right way to do so, but I'm not good enough in coding to know what's the right way. You can also tell me how to do it better or @pmrv or @liamhuber also has a good insight?

Reply:
Why not simply define it at the module level and use it in the property? If you do not want to use it outside of module, you could define it as private class `class _LammpsPairCoeff:` or the like.

Reply:
I'm ok with that. Since the efficiency hardly plays a role here, I just thought I'd put it within the function because it's even clearer that it's used only within the function.